src/numbers_royale.cpp

```diff
@@ -0,0 +1,188 @@
+#include "numbers_royale.h"
+#include "cpu.h"
+
+
+namespace {
+  // Helper functions
+  /**
+   * @brief Wait for user to press enter
+   * 
+   * @param message Message to display
+   */
+  void wait_for_enter(const char* message = "Press <enter> to continue... ") {
+    printf("%s", message);
+    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+  }
+
+  /**
+   * @brief Get the validated input object
+   * 
+   * @param is_valid lambda function to validate input
+   * @return int selected input after validation
+   */
+  int get_validated_input(const std::function<bool(int)>& is_valid) {
+    int input;
+    while (true) {
+      std::cin >> input;
+      if (std::cin.fail() || !is_valid(input)) {
+        std::cin.clear(); // clear the error flag
+        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // discard invalid input
+        std::cout << "Invalid input. Please try again: ";
+      } else {
+        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); // discard any extra input
+        return input;
+      }
+    }
+  }
+};
+
+
+// Numbers Royale Class
+NumbersRoyale::NumbersRoyale(): board_() {
+  GUI::clear_screen();
+}
+
+void NumbersRoyale::main_menu() {
+  while (true) {
+    GUI::print_header("NUMBERS ROYALE");
+    GUI::print_item("1. Play Game");
+    GUI::print_item("2. How to Play");
+    GUI::print_item("3. Quit");
+    GUI::print_border();
+    printf("Select option: ");
+    int option = get_validated_input([](int input) { 
+      return input >= MainMenuOption::PLAY_GAME && input <= MainMenuOption::QUIT; 
+    });
```

Comment:
```suggestion
    MainMenuOption option = static_cast<MainMenuOption>(get_validated_input());
```

Reply:
for if you want to use enum class

Reply:
your goal should be to try to get to stricter types when you can. Enums are pretty loose.

your stdin input is obviously untyped, and you're reading it as an int, which makes sense. From there you'll want to cast it into your type as soon as you can (here). And then you use it as that type. You want to process it as a raw int as little as possible

Reply:
Because `get_validated_input` takes in another lambda, would I just do some casting as well inside the input validation lambda?